lib/foreman/foreman_url_renderer.rb

```diff
@@ -23,6 +34,14 @@ def foreman_url(action = nil, params = {})
       host = self if @host.nil? && self.class < Host::Base
       template_proxy = host.try(:provision_interface).try(:subnet).try(:template_proxy)
 
+      # Parameters which must not be URL-encoded (e.g. iPXE synax ${xxx})
+      raw_string = ''
+      raw_params = params.delete(:unencoded)
```

Comment:
I was thinking about changing the signature of `foreman_url` to `def foreman_url(action = nil, params = {}, unescaped_params: {})` so it's done at the top level. That's easier to understand an document.

Reply:
I cannot do that, it will not work:

```ruby
>> def test(a = {}); puts a.inspect; end
=> :test
>> test(one: 1)
{:one=>1}
=> nil
>> def test(a = {}, b: {}); puts a.inspect, b.inspect; end
=> :test
>> test(one: 1)
Traceback (most recent call last):
        5: from /home/lzap/.rbenv/versions/2.7.1/bin/irb:23:in `<main>'
        4: from /home/lzap/.rbenv/versions/2.7.1/bin/irb:23:in `load'
        3: from /home/lzap/.rbenv/versions/2.7.1/lib/ruby/gems/2.7.0/gems/irb-1.2.3/exe/irb:11:in `<top (required)>'
        2: from (irb):18
        1: from (irb):17:in `test'
ArgumentError (unknown keyword: :one)
```

Reply:
Bleh, good point.

Reply:
hmph, mixing up args and kwargs is such a pain :crying_cat_face: